packages/host/app/services/realm-server.ts

```diff
@@ -173,7 +181,86 @@ export default class RealmServerService extends Service {
       });
   }
 
-  async fetchCreditInfo(): Promise<CreditInfo> {
+  async fetchCreditInfo() {
+    await this.fetchCreditInfoTask.perform();
+    await this.registerCreditInfoRefresher.perform();
+  }
+
+  get creditInfo() {
+    return this._creditInfo;
+  }
+
+  get fetchingCreditInfo() {
+    return this.fetchCreditInfoTask.isRunning;
+  }
+
+  async fetchCatalogRealms() {
+    if (this.catalogRealmURLs.length > 0) {
+      return;
+    }
+    let response = await this.network.authedFetch(
+      `${this.url.origin}/_catalog-realms`,
+    );
+    if (response.status !== 200) {
+      throw new Error(
+        `Failed to fetch public realms for realm server ${this.url.origin}: ${response.status}`,
+      );
+    }
+
+    let { data } = await response.json();
+    data.forEach((publicRealm: { id: string }) => {
+      if (!this.availableRealms.find((r) => r.url === publicRealm.id)) {
+        this.availableRealms.push({ type: 'catalog', url: publicRealm.id });
+      }
+    });
+    this.ready.fulfill();
+  }
+
+  get url() {
+    let url;
+    if (hostsOwnAssets) {
+      url = new URL(resolvedBaseRealmURL).origin;
+    } else {
+      url = globalThis.location.origin;
+    }
+
+    return new URL(url);
+  }
+
+  private registerCreditInfoRefresher = dropTask(async () => {
+    if (this.isCreditInfoRefresherRegistered) {
+      return;
+    }
+
+    if (!this.client || !this.sdk) {
+      throw new Error(
+        `Cannot register credit info refresher without matrix client or sdk`,
+      );
+    }
+
+    let sessionRoomId = await this.getSessionRoomId();
+    this.client.on(this.sdk.RoomEvent.Timeline, async (event: MatrixEvent) => {
+      await this.client?.decryptEventIfNeeded(event);
+      let content = event.getContent<{
+        msgtype: string;
+        body: string;
+      }>();
+      if (
+        event.getRoomId() !== sessionRoomId ||
+        event.getType() !== 'm.room.message' ||
+        content?.msgtype !== 'org.boxel.realm-server-event' ||
+        content?.body !== 'billing-notification'
+      ) {
+        return;
+      }
+
+      await this.fetchCreditInfoTask.perform();
+    });
```

Comment:
this is a weird place to do this binding, all the other matrix even bindings happen in `MatrixService.bindEventListeners()`. I would think we want to consolidate all this.

Reply:
Also, anytime you create a binding you should also be pairing it with an unbind--otherwise you can get into situations where you have multiple bindings to the same handler, like logging out and logging back it

Reply:
@habdelra I’ve updated the code. Instead of registering the refresher in `RealmServerService`, I added a call to `realmServerService.handleRealmServerEvent` within the `processDecryptedEvent` function in `MatrixService`. Since that function is already a listener for the `RoomEvent.Timeline` event, I thought it made sense. What do you think? If this approach looks fine, I’ll update the tests accordingly.